fix: add types to default and verify_token and Request __init__ based on comments in the source code.

wyattscarpenter opened the pull request:
Recently, I've been using this library with mypy strict mode, which doesn't like the fact that these functions are unannotated (they're the only functions I use, and thus the only ones mypy complains about to me). I'm happy to see that this project has py.typed, so these annotations should fix the problem!

arithmetic1728 commented:
Could you squash your commits, and change the commit message to "fix: add types to verify_token and request __init__"? The `conventionalcommits.org` test requires the commit message to start with the "fix:" prefix. Thanks!

wyattscarpenter commented:
Happy to!

wyattscarpenter committed:
fix: add types to verify_token and request __init__

Recently, I've been using this library with mypy strict mode, which doesn't like the fact that these functions are unannotated (they're the only functions I use, and thus the only ones mypy complains about to me). I'm happy to see that this project has py.typed, so these annotations should fix the problem!

wyattscarpenter committed:
Merge branch 'main' into patch-1

wyattscarpenter commented:
I'm hoping I can get this looked at in the near future. Also, it seems like (at least this time), it's the title of the PR that has to have fix: in it for conventionalcommits.org to pass, so  have done so.

wyattscarpenter committed:
type default based on comments

wyattscarpenter committed:
Update _default.py: wait actually I do want it to be Noneable, that's the point of default kind of.

wyattscarpenter committed:
Update _default.py: it's probably better to use the more-supported syntax

wyattscarpenter commented on google/auth/_default.py:
```diff
@@ -532,7 +536,12 @@ def _apply_quota_project_id(credentials, quota_project_id):
 
-def default(scopes=None, request=None, quota_project_id=None, default_scopes=None):
+def default(
+        scopes: Sequence[str],
```
Actually I was using the function without providing this parameter, so, you know, case in point. It's Optional now.

wyattscarpenter committed:
Update _default.py: attempt to avoid circular import

chalmerlowe committed:
Merge branch 'main' into patch-1

chalmerlowe committed:
Merge branch 'main' into patch-1

chalmerlowe committed:
Merge branch 'main' into patch-1

chalmerlowe commented on google/auth/_default.py:
```diff
@@ -588,7 +592,12 @@ def _apply_quota_project_id(credentials, quota_project_id):
-def default(scopes=None, request=None, quota_project_id=None, default_scopes=None):
+def default(
+        scopes: Optional[Sequence[str]] = None,
+        request: Optional[Request] = None,
```
```suggestion
        request: Optional["Request"] = None,
```

chalmerlowe reviewed (commented):
LGTM. Let's see how the tests turn out.
`Kokoro` and `Kokoro system-3.10` were failing due to circular imports. Hopefully we can nix that using the conditional.

chalmerlowe committed:
Update google/auth/_default.py

chalmerlowe committed:
Update google/auth/_default.py

chalmerlowe committed:
Update google/auth/_default.py

chalmerlowe committed:
Update google/auth/_default.py

chalmerlowe committed:
Update google/oauth2/id_token.py

chalmerlowe commented on google/auth/transport/requests.py:
```suggestion
import time
from typing import Optional
```

chalmerlowe commented on google/oauth2/id_token.py:
```diff
@@ -105,12 +108,12 @@ def _fetch_certs(request, certs_url):
-):
+    id_token: str | bytes,
+    request: transport.Request,
+    audience: str | list[str] | None = None,
```
```suggestion
    audience: Union[str, list[str], None] = None,
```

chalmerlowe committed:
Update google/oauth2/id_token.py

chalmerlowe committed:
Update google/oauth2/id_token.py

chalmerlowe committed:
Update google/auth/transport/requests.py

chalmerlowe committed:
Update google/auth/transport/requests.py

chalmerlowe committed:
Update google/auth/transport/requests.py

chalmerlowe commented on google/auth/transport/requests.py:
```diff
@@ -137,7 +138,7 @@ class Request(transport.Request):
     """
 
-    def __init__(self, session=None):
+    def __init__(self, session: [Optional[requests.Session]] = None) -> None:
```
```suggestion
    def __init__(self, session: Optional[requests.Session] = None) -> None:
```

chalmerlowe committed:
Update google/auth/transport/requests.py

chalmerlowe committed:
Merge branch 'main' into patch-1

chalmerlowe committed:
Update _default.py

Update to deal with linting issues.

chalmerlowe committed:
Update id_token.py

Updates to correct linting issues.

chalmerlowe committed:
Update _default.py

fixes to deal with linting and coverage errors.

chalmerlowe committed:
Apply suggestion from @chalmerlowe

missing a dadgum space character after the comment octothorpe. 🤯

chalmerlowe commented on google/auth/_default.py:
```diff
@@ -17,14 +17,20 @@
+
+if TYPE_CHECKING:  # pragma: NO COVER
+    from google.auth.credentials import Credentials
+    from google.auth.transport import Request
```
```suggestion
if TYPE_CHECKING:  # pragma: NO COVER
    # flake8 raises an error because these lines are not used in the code,
    # they are only used for type-hinting, hence the F401 pragma. 
    from google.auth.credentials import Credentials  # noqa: F401
    from google.auth.transport import Request  # noqa: F401
```

chalmerlowe committed:
Apply suggestion from @chalmerlowe

more linting stuff related to flake8 not liking our type hinting.

chalmerlowe committed:
Apply suggestion from @chalmerlowe

Updates order of imports and removes explanatory comments. flake8 does not like them, it does not match the expected pattern for this situation.

chalmerlowe commented:
This almost broke me.

For my own sake and anyone else who looks at all this commit hell:

* `blacken`, `flake8`, `mypy`, `coverage`, and `sphinx` all have rules for how things should be formatted, checked, and tested. Some of those rules differ by Python Runtime and some introduce subtle interactions that caused one OR more of these utilities to fail in ways that I did not expect.

I attempted to do the following:
* ensure that the type-hints were viable in under both Python 3.9 and 3.10+ since there was type hinting capabilities in 3.10 that was missing in 3.9
* avoid a circular import by adding an if condition related to type checking to control when/how specific libraries were imported i.e. only during type checking, not during runtime.
* comment the change so future me would remember why
* remove the comment because `flake8` has rigid rules about such type checking conditionals: it expects 3 lines between it and the last normal import statement: the import statement, a blank line, and the if statement. Comments added extra lines to the block and `flake8` was NOT happy.
* include full paths to the `Request` and `Credentials` type hints because sphinx saw multiple cross references to different `Request` and `Credential` types that come from different parts of the code and did not know how to cross reference them correctly.
* add `flake8` pragmas because `flake8` also didn't like the fact that the imports were not used in the code (`flake8` did not recognize that the if statement and code block were required for type checking... it just considered them unused)
* reorder the import statements to put the if conditional last, since `flake8` has rules about where such things can go
* add pragmas to ensure `coverage` won't flag missing unittest coverage for the if conditional (we chose not to unittest it since it is used solely by the type checker).
* ensure there are two spaces in front of the pragma because blacken requires two spaces:
  * `if TYPE_CHECKING:  # pragma: NO COVER`
  versus
  * `if TYPE_CHECKING: # pragma: NO COVER`
* add several miscellaneous commits and tests to try and discover what was happening with all of these intermingled rules, etc.

chalmerlowe reviewed (approved):
LGTM

chalmerlowe committed:
Merge branch 'main' into patch-1

chalmerlowe committed:
Merge branch 'main' into patch-1

wyattscarpenter commented:
Thanks!
